core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```
@@ -0,0 +1,398 @@
+        case None =>
+          (Errors.INVALID_PID_MAPPING, null)
+
+        case Some(metadata) =>
+          // generate the new transaction metadata with added partitions
+          metadata synchronized {
+            if (metadata.pid != pid) {
+              (Errors.INVALID_PID_MAPPING, null)
+            } else if (metadata.producerEpoch != epoch) {
+              (Errors.INVALID_PRODUCER_EPOCH, null)
+            } else if (metadata.pendingState.isDefined) {
+              // return a retriable exception to let the client backoff and retry
+              (Errors.COORDINATOR_LOAD_IN_PROGRESS, null)
+            } else if (metadata.state != Empty && metadata.state != Ongoing) {
+              (Errors.INVALID_TXN_STATE, null)
+            } else if (partitions.subsetOf(metadata.topicPartitions)) {
+              // this is an optimization: if the partitions are already in the metadata reply OK immediately
+              (Errors.NONE, null)
+            } else {
+              val now = time.milliseconds()
+              val newMetadata = new TransactionMetadata(pid,
+                epoch,
+                metadata.txnTimeoutMs,
+                Ongoing,
+                metadata.topicPartitions ++ partitions,
+                now,
+                now)
+              metadata.prepareTransitionTo(Ongoing)
+              (Errors.NONE, newMetadata)
+            }
+          }
+      }
+
+      if (newMetadata != null) {
+        txnManager.appendTransactionToLog(transactionalId, newMetadata, responseCallback)
+      } else {
+        responseCallback(error)
+      }
+    }
+  }
+
+  def handleTxnImmigration(transactionStateTopicPartitionId: Int, coordinatorEpoch: Int) {
+    txnManager.loadTransactionsForPartition(transactionStateTopicPartitionId, coordinatorEpoch)
+  }
+
+  def handleTxnEmigration(transactionStateTopicPartitionId: Int) {
+    txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
```

This can be tricky to handle completely. (1) Should we also cancel any ongoing loading due to the previous immigration? (2) There could be outstanding transactional requests (e.g., initPid) waiting in producer purgatory after the log append call. Ideally, we should mark the coordinator as not available, trigger a check on requests associated with the coordinator's partition in the purgatory so that they can responds with a coodinator_not_available error.

---

@junrao i think we can handle (1) by:
1. in `removeTransactionsForPartition` we remove the partition from `loadingPartitions`
2. in `loadTransactionMetatdata` we only perform the loop when `loadingPartitions.contains(partition)

I have to think a bit more about 2

---

@junrao i'm not sure i understand (2). When you refer to there being outstanding requests in the producer purgatory is that w.r.t `replicaManager.appendRecords(...)`?

---

Yes, I was referring to delayed events in the producer purgatory after the replicaManager.appendRecords(...) call. We can trigger a check of those delayed events and let them error out (since the check will find out that the broker is no longer the coordinator).

---

@junrao is calling `delayedProducePurgatory.checkAndComplete(topicPartitionOperationKey)` the correct thing to do in this case?